src/smstateen.adoc

```diff
@@ -0,0 +1,335 @@
+A hypervisor may need this control over
+accesses to the `sstateen` CSRs if it ever must emulate for a virtual machine an
+extension that is supposed to be affected by a bit in an `sstateen` CSR. (Even if
+such emulation is uncommon, it should not be excluded.) Machine-level software
+needs identical control to be able to emulate the hypervisor extension. (That
+is, machine level needs control over accesses to the supervisor-level `sstateen`
+CSRs in order to emulate the `hstateen` CSRs, which have such control.)
+
+Bit 63 of each `mstateen` CSR may be read-only zero only if the hypervisor
+extension is not implemented and the matching supervisor-level `sstateen` CSR is
+all read-only zeros. In that case, machine-level software should emulate
+attempts to access the affected `sstateen` CSR from S-mode, ignoring writes and
+returning zero for reads. Bit 63 of each `hstateen` CSR is always writable (not
+read-only).
+
+Initially, the following bits are defined in `mstateen0`,
+`hstateen0`, and `sstateen0`:
+
+bit 0 - Custom state
+
+bit 1 - `fcsr` for Zfinx and related extensions (Zdinx, etc.)
+
+Bit 0 controls access to any and all custom state.
+
+(Bit 0 of these registers is not custom state itself; it is a standard field of
+a standard CSR, either `mstateen0`, `hstateen0`, or `sstateen0`. The requirements
+that non-standard extensions must meet to be _conforming_ are not relaxed due
+solely to changes in the value of this bit. In particular, if software sets
+this bit but does not execute any custom instructions or access any custom
+state, the software must continue to execute as specified by all relevant
+RISC-V standards, or the hardware is not standard-conforming.)
+
+Bit 1 applies only for the case when floating-point instructions operate on `x`
+registers instead of `f` registers. Whenever `misa`.F = 1, bit 1 of `mstateen0` is
+read-only zero (and hence read-only zero in `hstateen0` and `sstateen0` too). For
+convenience, when the `stateen` CSRs are implemented and `misa`.F = 0, then if bit
+1 of a controlling `stateen0` CSR is zero, _all_ floating-point instructions
+cause an illegal instruction trap (or virtual instruction trap, if relevant),
+as though they all access `fcsr`, regardless of whether they really do.
+
+In addition to the bits listed above for user-accessible state, the following
+are also defined initially for `mstateen0`:
+
+bit 57 - `hcontext`, `scontext`
+
+bits 60:58 - Reserved for the RISC-V Advanced Interrupt Architecture
+
+bit 61 - Reserved for possible `henvcfg2`/`henvcfg2h`, `senvcfg2`
+
+bit 62 - `henvcfg`/`henvcfgh`, `senvcfg`
+
+bit 63 - `hstateen0`/`hstateen0h`, `sstateen0`
+
+The bits defined initially for `hstateen0` are the same as those for `mstateen0`
+except applying only to state that is accessible in VS-mode:
+
+bit 57 - `scontext`
+
+bits 60:58 - Reserved for the RISC-V Advanced Interrupt Architecture
+
+bit 61 - Reserved for a possible `senvcfg2`
+
+bit 62 - `senvcfg`
+
+bit 63 - `sstateen0`
```

Comment:
Should be converted into a register layout diagram.

Reply:
Prior to merge or can this be done post merge?  If pre-merge, someone will have to do it.

Reply:
@ved-rivos so what do you think, can the register layout diagram be added post chapter merge?